# Exploration of Amazon consumer products reviews
## by Kyle McMillan

## Preliminary Wrangling

> This investigation explores the Amazon US marketplace reviews dataset containing approximately 7 million unique reviews.

In [92]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import re
import random
random.seed(42)

%matplotlib inline

In [93]:
#Data is loaded from a TSV file. Bad lines are dropped as fixing these manually will take too much time 
#for the low number of lines with errors.
reviews=pd.read_csv(r"amazon_reviews_multilingual_US_v1_00.tsv", sep="\t", error_bad_lines=False, header=0)

b'Skipping line 3231472: expected 15 fields, saw 22\n'
b'Skipping line 3509762: expected 15 fields, saw 22\n'
b'Skipping line 4018793: expected 15 fields, saw 22\n'
b'Skipping line 4280173: expected 15 fields, saw 22\nSkipping line 4290596: expected 15 fields, saw 22\n'
b'Skipping line 4331421: expected 15 fields, saw 22\nSkipping line 4340267: expected 15 fields, saw 22\nSkipping line 4341665: expected 15 fields, saw 22\nSkipping line 4386155: expected 15 fields, saw 22\nSkipping line 4388098: expected 15 fields, saw 22\n'
b'Skipping line 4408027: expected 15 fields, saw 22\nSkipping line 4442615: expected 15 fields, saw 22\n'
b'Skipping line 4519623: expected 15 fields, saw 22\n'
b'Skipping line 4525797: expected 15 fields, saw 22\nSkipping line 4543519: expected 15 fields, saw 22\n'
b'Skipping line 4587726: expected 15 fields, saw 22\nSkipping line 4589301: expected 15 fields, saw 22\nSkipping line 4634393: expected 15 fields, saw 22\n'
b'Skipping line 4666168: expected 15 fields, s

In [94]:
#View the unique items in the product category.
reviews.product_category.unique()

array(['Books', 'Music', 'Video', 'Video DVD', 'Toys', 'Tools',
       'Office Products', 'Video Games', 'Software',
       'Digital_Music_Purchase', 'Home Entertainment', 'Electronics',
       'Digital_Ebook_Purchase', 'Digital_Video_Download', 'Kitchen',
       'Camera', 'Outdoors', 'Musical Instruments', 'Sports', 'Watches',
       'PC', 'Home', 'Wireless', 'Beauty', 'Baby', 'Home Improvement',
       'Apparel', 'Shoes', 'Lawn and Garden', 'Mobile_Electronics',
       'Health & Personal Care', 'Grocery', 'Luggage',
       'Personal_Care_Appliances', 'Automotive', 'Mobile_Apps',
       'Furniture', '2012-12-22', 'Pet Products'], dtype=object)

In [95]:
#View the data that has a product category of a date.
reviews.query("product_category == '2012-12-22'")

marketplace  customer_id       review_id  product_id  product_parent  \
1852794          US     49422747  R2T9JZNQ996WRC  1568652240       569473707   

                                             product_title product_category  \
1852794  Emma (Large Print)\tBooks\t5\t0\t0\tN\tY\tFoll...       2012-12-22   

         star_rating  helpful_votes  total_votes vine verified_purchase  \
1852794          NaN            NaN          NaN  NaN               NaN   

        review_headline review_body review_date  
1852794             NaN         NaN         NaN

In [96]:
#View the data string in the product title
list(reviews[reviews.review_id=="R2T9JZNQ996WRC"].product_title)

['Emma (Large Print)\tBooks\t5\t0\t0\tN\tY\tFollows Austen\'s book closley\tI have the movie with Ms Paltrow starring, so I bought the book to see how it matched. Very well done.\t2012-12-22\nUS\t7209000\tR1MZOZGBLVKYVP\tB003E8P9G0\t446279348\tThe Kane Chronicles, Book One: The Red Pyramid\tDigital_Ebook_Purchase\t5\t0\t0\tN\tN\tso....\tmy brother read this book and he was instantly addicted to it so i started reading it on my kindle and i goten adddicted to it too. i would say that it is a must read.Questions or Comments my kindle email is  haili@kindle.com\t2012-12-22\nUS\t49422747\tR31UI3EECPWNVA\tB002EWD0I6\t647475881\tLark Rise to Candleford: Season 1\tVideo DVD\t4\t0\t0\tN\tY\tGood\tI thought this would be more of theprevious characters of Cranford, but not to be.  This was good enough for me to order the 2nd season.\t2012-12-22\nUS\t7209000\tR2UXMYNK8AEEVR\tB005CRQ4GU\t632021160\tThe Third Wheel (Diary of a Wimpy Kid, Book 7)\tDigital_Ebook_Purchase\t5\t1\t2\tN\tN\tso....\tI LOV

In [97]:
#Drop this line as it appears to have been parsed wrong because of the extra "/t" in the title column and there are a lot
#of extra reviews in this title string.
reviews.drop(reviews.loc[reviews.product_category.isin(['2012-12-22'])].index, inplace=True)
reviews.product_category.unique()

array(['Books', 'Music', 'Video', 'Video DVD', 'Toys', 'Tools',
       'Office Products', 'Video Games', 'Software',
       'Digital_Music_Purchase', 'Home Entertainment', 'Electronics',
       'Digital_Ebook_Purchase', 'Digital_Video_Download', 'Kitchen',
       'Camera', 'Outdoors', 'Musical Instruments', 'Sports', 'Watches',
       'PC', 'Home', 'Wireless', 'Beauty', 'Baby', 'Home Improvement',
       'Apparel', 'Shoes', 'Lawn and Garden', 'Mobile_Electronics',
       'Health & Personal Care', 'Grocery', 'Luggage',
       'Personal_Care_Appliances', 'Automotive', 'Mobile_Apps',
       'Furniture', 'Pet Products'], dtype=object)

In [98]:
print(reviews.info(null_counts=True ))

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6900885 entries, 0 to 6900885
Data columns (total 15 columns):
marketplace          6900885 non-null object
customer_id          6900885 non-null int64
review_id            6900885 non-null object
product_id           6900885 non-null object
product_parent       6900885 non-null int64
product_title        6900885 non-null object
product_category     6900885 non-null object
star_rating          6900885 non-null float64
helpful_votes        6900885 non-null float64
total_votes          6900885 non-null float64
vine                 6900885 non-null object
verified_purchase    6900885 non-null object
review_headline      6900811 non-null object
review_body          6900810 non-null object
review_date          6900562 non-null object
dtypes: float64(3), int64(2), object(10)
memory usage: 842.4+ MB
None


In [99]:
#Investigate the reason that a few reviews have NaN for a date
dateless = reviews.loc[reviews['review_date'].isin(reviews.review_date.dropna().unique())==False]
dateless.head()

marketplace  customer_id       review_id  product_id  product_parent  \
142642          US     52767614   RJBEAQJ92LS61  B00003CXDG       765637830   
186564          US     50374272  R2LXX6V7B1PICZ  B00005OB0A       321992753   
217791          US     42627253  R1CRCYSML85MB4  B00005JKZQ       167276143   
235044          US     52442862  R1YZIXBH6AVF4E  B00000JMQC        79142420   
239621          US     47692344  R12KVFMZYPXNF6  B00003CWL6       304141589   

                                     product_title product_category  \
142642  Mission: Impossible 2 (Widescreen Edition)        Video DVD   
186564                                       Fever            Music   
217791                                    Showtime        Video DVD   
235044                          Return to Oz [VHS]            Video   
239621                      American Beauty (1999)        Video DVD   

        star_rating  helpful_votes  total_votes vine verified_purchase  \
142642          2.0            0.0          0.0    N                 N   
186564          5.0            0.0          0.0    N                 N   
217791          5.0            5.0          6.0    N                 N   
235044          5.0            6.0          7.0    N                 N   
239621          5.0            1.0          3.0    N                 N   

                                          review_headline review_body  \
142642  Technology is not always beneficial -\tespecia...  2001-03-08   
186564  The music you were playing really blew my mind...  2001-10-25   
217791  It's............SHOWTIME!!!!!\tShowtime is the...  2002-03-16   
235044  The Wizard of Oz it is not--should not be!\tFo...  2002-06-10   
239621  It's just STUFF!!!\t\\American Beauty\\" is in...  2002-06-30   

       review_date  
142642         NaN  
186564         NaN  
217791         NaN  
235044         NaN  
239621         NaN

In [100]:
#It seems there was an issue during the reading of the TSV file and the review titles used a tab space. Where read_csv was
#looking for a "\t" delimiter

#Investigate the titles for each of the reviews to see if there is any thing unusual
#As this is a very large and long list, a random sample of 5 were chosen.
DL_random_list = random.sample(list(dateless.review_id), 5)
for i in DL_random_list:
    print(re.split("\t", list(dateless[dateless.review_id==i].review_headline)[0])[0])

The Lone survivor
My teacher said I'm a loser, I give a f**k if you feel me..
Love the boys,  but not this movie
What did you expect?.....Lord Tennyson?
Oldie, but goodie


In [101]:
#Investigate the body of the dateless reviews to see if there is anything unusual.
for i in DL_random_list:
    print(re.split("\t", list(dateless[dateless.review_id==i].review_headline)[0])[1:])

["[[ASIN:0316067601 Lone Survivor: The Eyewitness Account of Operation Redwing and the Lost Heroes of SEAL Team 10]]<br /><br />Excellent narration. Inspiring, patriotic, and courageous. Also, one lucky SEAL. As an American, I am honored, and humbled by his 'call to service\\\\.<br />Bravo."]
['.. I\'m gonna follow my heart.\\\\ <BR>My favorite line on this album off of the highest point on this album, \\\\"Get Em High,\\\\" featuring epic rappers Talib Kweli and Common. I loved it because it influences me and inspires me, but after learning that this was the genius that produced five star tracks such as Poppa Was A Playa by Nas and Izzo by Jay-Z, I wanted to pick up this solely for the beats. I wasn\'t as interested in hearing him as a rapper, because I hadn\'t even heard Through the Wire or All Falls Down at the time. But, I wasn\'t disappointed at all, this guy is one of the best \\\\"new artist\\\\" rappers I\'ve heard in a long time. This guy knows how to write intellectual flows 

In [102]:
#It seems like there are a lot of other reviews mixed into the body of some of the dateless reviews.
#As there are a very large number of total reviews, and wrangling the data to sort out the 300 or so that need to be fixed,
#I have decided to drop these rows.
reviews.drop(dateless.index, inplace=True)
reviews.loc[reviews['review_date'].isin(reviews.review_date.dropna().unique())==False]

Empty DataFrame
Columns: [marketplace, customer_id, review_id, product_id, product_parent, product_title, product_category, star_rating, helpful_votes, total_votes, vine, verified_purchase, review_headline, review_body, review_date]
Index: []

In [103]:
#Investigate the reviews where there is no headline to see if there is anything unusual.
reviews.loc[reviews['review_headline'].isin(reviews.review_headline.dropna().unique())==False]

marketplace  customer_id       review_id  product_id  product_parent  \
144995           US     51263804   RRR6EK045G5YL  630438551X       212477251   
274671           US     34490235  R25MC0QXPV5CRZ  6305949980       664817538   
373447           US     23532323  R1MNPI81JJ62NI  B0000DD7NL       868764779   
450489           US     43058957   RIHK7J9KBB62V  B00005V3Z4       681790048   
515646           US     34215909  R1HAKDWL5X86B9  B00014NE62        59241661   
522262           US     36882963  R2VEG37XSHHQMO  B0002Y4TTC       773850099   
546262           US     24610674  R3NODAYWJSH67Y  B000BM6AVA       828472430   
550148           US     16196376  R1DUI7MJ3FPONG  B0001YRVN4       745272701   
567259           US     22947656  R1ABIC2ULD2TCB  B000AP2ZDK       577609284   
568389           US     42224700  R3QMAWGAVYTR9F  B00001U0E1       475263130   
586782           US     31454572  R1WWDBIRZCTWC7  B00005IC2H       289287264   
629660           US     10076686  R1BJZCW8XVWANT  0618680004       362925721   
657901           US     16775323   RA8S2CO9Q1C4A  B000IOM1RI       624386805   
983962           US     22059507  R1CW7KH3BNYUR8  0446618306       286127559   
1016754          US     20130602  R2C6PGGPYXBS06  B000HC2LI0       820535880   
1297103          US     18207773  R28MXJ4COFS7L5  B003YOZNBK       886437736   
1512037          US     21090649  R1S0I91296OAUX  B000RGSOR8       744692957   
1567945          US     19063708  R26ROLEQOECQV6  B007BVJSLS       169791471   
1720187          US     50830744  R1BQUUT1ADPIEV  B004MC8CA2       937221751   
1750846          US     48437078  R37C3588L947A9  B004SBX112        51729933   
1750848          US     48437078  R3C3XSP3XQQU5Q  B000000OQY       220149902   
1750850          US     48437078   RIBS8E10HG295  B0063IH60K       414486782   
1750852          US     48437078  R3LEI31HI10DMP  0790729652       291187518   
1750854          US     48437078  R1R4HAL6ZT34XQ  B00004Z4WR       208987230   
1750856          US     48437078  R1XRWGVO7LJZU1  B00001NFCY       354785034   
1750858          US     48437078  R2Z78Q9HPIFVSM  B004JZBGWI         3534159   
1750860          US     48437078   RC7NNCFA97EAY  B004SJ3BCI       798420614   
1750862          US     48437078   R6PU5QMNUF28V  B004SJ3AXI        69013933   
1772839          US     42011346  R11V3LSWKBNNY8  B0091LS07E       312605089   
1846430          US     38563223  R2RFT4TI7Q4BLD  B0063IH60K       414486782   
...             ...          ...             ...         ...             ...   
3078345          US     19690188   R49AR6KWA3HQN  0385535147       273127372   
3136342          US     17153687  R3NS0U2DUSQEBJ  076781326X       869564928   
3286342          US     14019317   RL1J9U0SMNKNZ  B0067AC5KA       309710589   
3352344          US     26510003   RAE9HY2T4YFCW  B00BWBHIUG       835549535   
3362918          US     42218628  R1ELYSCQDQ0F91  B0052FYPJK       891672634   
3365536          US     26650224  R3PB45NO1W6JIF  B003G4W49C       401439625   
3549947          US     46671268  R2ZXKTGQJT4JA6  B00594XOZU       191341525   
3644880          US     12920423   ROUL0UB9GF3R6  B00F8H0LM2       791547247   
3852718          US     16211327  R245GNUXWH3SQB  B0058YPGSY       659245091   
3895915          US     21265913   R2TUGY7YT8IHX  B0094BB4TW       569156287   
3951069          US     17210931   RM256DI7LV56D  B00DOPEGAS       948324162   
4099711          US      7682720  R1WKAOELTTKLON  B00J8NDGOM       739373186   
4109055          US     14736916  R1DZ6JT3V7UZES  B008476HBM       503057596   
4179320          US     43821421  R2IN7GSZQ4L5FO  B002ZCY7SW       844845858   
4197447          US     12151325   R4512LFXE0A3O  B009MF56GC       490387818   
4226306          US     16607364  R11KN1Z67T50Y7  B00IQDFBN4       645171913   
4634477          US      6940513  R2AEG5SFQT085K  B00IG2DOKM       784083581   
4677659          US     45065979   RWHZQL7YE75HU  1401924751       965620385   
5066294 

In [104]:
#Investigate the reviews where there is no body to see if there is anything unusual.
reviews.loc[reviews['review_body'].isin(reviews.review_body.dropna().unique())==False]

marketplace  customer_id       review_id  product_id  product_parent  \
4362503          US     44409249  R28HX7V41NGPGD  B00IFMHZ58       731881208   
4364915          US     34483539  R1KRVATE3RM685  0545615402       160159764   
4492736          US     14326822  R3QYW87GDRRNAF  B004S82OAE       235122690   
4646674          US     20596625  R12WMFWQ96MK9E  B00HPYMVD8       108613172   
4867197          US     11236181  R1ZS0RW3SLED36  B0096YJDNQ       454110113   
4967822          US      3152188  R1RGMJWY7KCMFF  B003MYYJD0       120446899   
4990204          US      8149238  R221BRF51GRMNT  B004GJDQT8        36653526   
5155825          US     33444040  R2XT5Y4GSMFGBJ  B000LXQVA4       535123469   
5166126          US     44256154   RK1UNP8GPEEKV  B002IIY6X4       795348432   
5190776          US     32311114  R1Q33KCFX6EZKX  B001XVD21Y       818712447   
5293803          US     12612709  R1Y50GT2992THK  0812979680       632326343   
5343709          US     46753638  R37ZVWUS38O2F6  0141393394       655343360   
5470188          US      4718213  R3SHXDX8DDF0OS  B0094BB4TW       569156287   
5500826          US     16128297  R1KPKBNXZO4JB9  B000KJT8IG       101502671   
5562798          US     52508298  R14HQFJW463FV5  B00J4S962W       601447693   
5609738          US      2479412   RCTH5Z82KPWP3  B00I3ULQQ6       975900063   
5652410          US      9187235  R1KTOMNUZOG99S  B002JVWRAO       366071566   
5657378          US      1117070  R3QZ7UPF23932S  B00C3JU7D6       137422612   
5667952          US     39253932  R12LWUURDP3450  B0041L8ZRY       761482228   
5669881          US      6040618  R2PM5VYLWVWHWV  B00MU1YF9Y       195017722   
5693766          US      7294686   R1FVFAUXEED5P  B009KS4XRO       490050721   
5732816          US     34018263   RNRIXD0YYBH7W  B008Y78AO4       575221740   
5746533          US      3353098  R3MG8S8JEG25P1  B00H71PBQ8       530546487   
5794502          US      4281799  R2FIXXX9TVX0BE  0147515025       712432151   
5818898          US      6821186   R7CX6EZDFRRKW  B00OKYQZDI       155730994   
5850825          US       828567  R1C35AXA9ZFF5R  B004GJDQT8        36653526   
5876128          US     24666014  R1V7GEAAZUCGT3  B002QYNIHU       704929602   
5879733          US       433520  R3IEUV6QNPR9X5  B009LDXQGW       972743481   
5967737          US     42685483  R2SWXJC80WJZOS  B007NLT6J0        10877377   
5976553          US     11765899   RIN2AP6DBV65R  B000LX00RY       415826823   
...             ...          ...             ...         ...             ...   
6495260          US     43256067   RT4N4DLKLQNET  1101946342       474654167   
6542987          US     32440033   RQNGW7WALQ3E2  B00I2TV3NO       290976617   
6546219          US       827493   RX2L20OBQDIHY  B0006AAS5G         1348996   
6568161          US     34107046   RBL5JB8JZF44D  B00QVZ1VXI       471257825   
6590914          US      1020177   RF2GPS9GE4D4Z  1101946342       474654167   
6605174          US       519656   RI7F3Z940RLZZ  B004GJDQT8        36653526   
6614621          US       469830  R35GKQG0X73G6J  B004GJDQT8        36653526   
6628037          US       563316  R24J6E79DICW96  B00AEFY0F2       727294557   
6633991          US     51011399   RS5RYBBFF8DEF  0061766089       244828109   
6634261          US     41382231   RDYL1KNEKACEY  B004FRX0MY       473764126   
6639783          US     11190013  R1PC11OVU7ACGD  B008XG1X18       752212580   
6651579          US     38484571   R74VTHE48J4IQ  B0028IDXDS       450475749   
6653694          US      2564645   RFY7DV76OTDIP  B006TLZNDI       363585395   
6656032          US     21965134  R1KFHK1M0IQE1Z  B00000F6LQ       401858789   
6677378          US     10116686   R7PEQJRZX8KCN  B00JB3MVCW       109782764   
6688326          US      3179150  R3VUDCXL4EBUVO  B00ENYKBD0       549659889   
6691748          US       261583   RL5YKLA90PEG8  B004GJDQT8        36653526   
6709065          US     26966424  R1WUQJ3PZ8T8XD  0763619612       609689991   
6764942 

Looking at the 2 sub tables of the bodyless and titleless reviews, it seems like there were no parsing errors with these but rather people simply did not write anything in these fields when writing their review.  
As such, I will keep these as they are and not remove them.  

There were a few lines that were dropped, these were to save time with this dataset. In total there are around 7 million reviews in this data set, and some need to be fixed manually. These rows of data of approximatley 1000 reviews were dropped as the final table has slightly more than 6.9Million rows.

In [105]:
#As this dataset is only for the US marketplace and all of these values are "US" this column is dropped.
reviews.drop("marketplace", inplace=True, axis=1)

In [106]:
#Check for duplicate rows.
sum(reviews.duplicated())

0

In [107]:
#Convert columns to the correct data type.
reviews['review_date']=pd.to_datetime(reviews.review_date)
reviews['vine'] = reviews.vine.map({'N':False, 'Y':True}).astype(bool)
reviews['star_rating']=pd.to_numeric(reviews.star_rating, downcast="integer")
reviews['helpful_votes']=pd.to_numeric(reviews.helpful_votes, downcast="integer")
reviews['total_votes']=pd.to_numeric(reviews.total_votes, downcast="integer")
reviews['verified_purchase'] = reviews.verified_purchase.map({'N':False, 'Y':True}).astype(bool)
reviews['customer_id'] = reviews.customer_id.astype(str)
reviews['product_parent'] = reviews.product_parent.astype(str)

In [108]:
#High level overview of the data shape and composition.
print(reviews.shape)
print(reviews.info(null_counts=True ))
print(reviews.head(10))

(6900562, 14)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 6900562 entries, 0 to 6900885
Data columns (total 14 columns):
customer_id          6900562 non-null object
review_id            6900562 non-null object
product_id           6900562 non-null object
product_parent       6900562 non-null object
product_title        6900562 non-null object
product_category     6900562 non-null object
star_rating          6900562 non-null int8
helpful_votes        6900562 non-null int16
total_votes          6900562 non-null int16
vine                 6900562 non-null bool
verified_purchase    6900562 non-null bool
review_headline      6900488 non-null object
review_body          6900487 non-null object
review_date          6900562 non-null datetime64[ns]
dtypes: bool(2), datetime64[ns](1), int16(2), int8(1), object(8)
memory usage: 572.5+ MB
None
  customer_id       review_id  product_id product_parent  \
0    53096384   R63J84G1LOX6R  1563890119      763187671   
1    53096399  R1BALOA11Z06MT

In [109]:
#Save a sample of the table to upload to github, as the original file is far to large to be uploaded.
reviews.sample(n=10000, random_state=42).to_csv(r'amazon_reviews_sample.csv', index=False)

In [110]:
#Satistics of review ratings
print(reviews[["star_rating","helpful_votes","total_votes"]].describe())

        star_rating  helpful_votes   total_votes
count  6.900562e+06   6.900562e+06  6.900562e+06
mean   4.306589e+00   2.044490e+00  3.251608e+00
std    1.146197e+00   3.184562e+01  3.634051e+01
min    1.000000e+00   0.000000e+00  0.000000e+00
25%    4.000000e+00   0.000000e+00  0.000000e+00
50%    5.000000e+00   0.000000e+00  0.000000e+00
75%    5.000000e+00   1.000000e+00  2.000000e+00
max    5.000000e+00   2.755000e+04  2.872700e+04


In [111]:
print("Number of unique customers:", len(reviews.customer_id.unique()))
print("Number of unique goods:", len(reviews.product_parent.unique()))
print("Number of unique catergories:", len(reviews.product_category.unique()))

Number of unique customers: 4097672
Number of unique goods: 16309
Number of unique catergories: 38


In [112]:
x=reviews.customer_id.value_counts()
y=x[x >5].index.tolist()
reviews.loc[reviews['customer_id'].isin(y)]

customer_id       review_id  product_id product_parent  \
34         53090280  R1R1V2QRQ87Q92  0440178002      191357398   
52         52977241   RULWKAINTXAUR  0393314804      232683594   
55         53090280  R200N7BA6G2YXT  0440178002      191357398   
69         53093818  R13OOAVKRQIFSO  0812548094      939049888   
94         52748904  R1F2IGEGT8M6T4  0553346687       33679552   
184        53051790   RXI5DPIRSBKVF  0812550706      401439625   
212        52636801  R3PQQG9U109PP1  0312857063      939049888   
315        52977842  R28F1BBT2O57KQ  0440213525      282981210   
490        52885970  R3638HA7J3999A  0345315227      284653366   
589        53067292  R1HZE2HEHGVHRR  044651652X      879659950   
593        53067292  R35XSEA6BCBCK6  0517702967      922959758   
633        53059087   RPDBHO5HK0MV1  0684874350      698161422   
856        52859652   RMZRQ4TH9U1PG  0553278398      525478656   
858        52859652  R2OC78NS6X59KZ  0812550706      401439625   
890        52962111  R1EZOYG9XQAHNV  044651506X      348227423   
917        51376733  R244LMLBGA1DIF  0451452739      379678595   
945        53027010  R33HOBM9MTAEP4  0812524586      401439625   
963        52986477  R3KEE01CX2A1KD  0930289234      777411096   
964        52064040  R2Y7V5GUZQTZ0M  0385472951      999720718   
974        52859652  R132Q2VKSJE5DT  0671004107      860650224   
978        52687118  R242HA2EQ47UBK  B000002UB3       96770875   
980        52687118   R70TQ5FLN5BG0  B000002UA9      243110428   
982        52687118  R1582E2C5XI3I8  B000002UB6       96614327   
985        52687118   RPRR257OR6L1K  B000002UAC      619776604   
987        52687118  R1321BD7LFFILD  B000002UAI      565955686   
988        52687118  R2PBJPXXIKXIVS  B000002UAF       16614211   
1017       53053685  R2YYZ795WFAZCF  0886461251      458864358   
1042       52687118  R1867UBYOPSEV0  B000002UAO        4895975   
1044       52687118   R9GOK37HULXDB  B000002UAR      153597720   
1092       53027010  R314PO3XA5OQIF  0783815484      781565906   
...             ...             ...         ...            ...   
6900574    52060656  R1PI1JYKXJE3YI  B004SIIBGU       30996242   
6900576    52060656    RMKT9891F03N  B004SOR4H6      255395670   
6900590    47217281  R3VZKW6MVN6WAI  B005HEJ2DK      533733733   
6900596    46126645  R13BR7XSL49QWN  B00AEFXCG0      506477919   
6900598    46126645  R3U4BJ8S5IJ5JV  B000FGGE7C      836017697   
6900623    23408661  R2VYF3W8U2ON1B  0307277453      883454197   
6900651    26872631  R3FFZY275K788V  B000002P72       41938886   
6900686      898423  R3BZ4Y1H11GPHO  B0094BB4TW      569156287   
6900688      898423  R1ZW8CV9I01X3V  B00HGJ18C2      326848133   
6900690      898423  R2EPVNDU2T7OS4  B0064X7FVE      140639925   
6900692      898423   R8GWUM64RNZXP  B008XG1X18      752212580   
6900706    52624155   RUR34KHR3TE7H  B005LAIHSG      206450753   
6900707    11672472  R1WJDLDDJUWDHD  B000HL0EEE      112960517   
6900710    51617406  R1YVNC08UXVWTT  B00WP9HSJU      727359307   
6900712    51131029  R2DDOC54DLWXJ1  B008NW1M4K      691626223   
6900716    48495146  R2INBFNATQ1EGJ  B00USC3SUY      153919204   
6900726    45121976  R1OS6UMQMX92BB  B000005RUV      298662366   
6900748    28459589  R1UCCUYC39UKFW  B005HED11Y      583611354   
6900761    51896378  R2FWT8HRP8JD5W  B008K8VDPU      481677245   
6900763    51018594  R1O51DKAPM1VTC  B00JCW0BCY      451025339   
6900764    18852188  R2JK8I43ETJGZU  B00L9B7IKE      627793267   
6900767    49881820   R3LOC1D2XDU2S  B008D28HYS      608399083   
6900769    49240092  R3L9DUFZYF9TFM  B004SD5GZ4      826010808   
6900771    49240092   RRABA6WF1FGFZ  B00P42WROG      756230011   
6900796    27624141  R3CDAAJM44OEM1  B0089LOG02        6213846   
6900814    23439210  R3O6QRXCQ3K1SL  B00QQW1VJA      647130552   
6900833    13920401  R3VBBEUI32M6WU  B00MFDOYUC      656086450   
6900834    20724599  R3MRKBOMW1S3I7  B00633W614      464759152   
6900848    15994446  R3T501CJCQB8F8  B000025EW7 

In [113]:
reviews.groupby('customer_id').count() 

review_id  product_id  product_parent  product_title  \
customer_id                                                         
100000               1           1               1              1   
10000030             1           1               1              1   
10000031             1           1               1              1   
10000037             3           3               3              3   
10000040             1           1               1              1   
10000049             1           1               1              1   
10000059             1           1               1              1   
10000064             2           2               2              2   
10000087             2           2               2              2   
10000097             1           1               1              1   
10000105             1           1               1              1   
10000109             2           2               2              2   
10000116             1           1               1              1   
10000119             1           1               1              1   
10000128             1           1               1              1   
10000138             1           1               1              1   
10000153             1           1               1              1   
10000162             1           1               1              1   
10000166            43          43              43             43   
10000205             1           1               1              1   
10000206             1           1               1              1   
10000223             1           1               1              1   
10000229             1           1               1              1   
10000235             1           1               1              1   
10000247             1           1               1              1   
10000271             4           4               4              4   
10000278             1           1               1              1   
10000280             1           1               1              1   
10000282             1           1               1              1   
10000285             1           1               1              1   
...                ...         ...             ...            ...   
9999739              1           1               1              1   
9999741              1           1               1              1   
9999744              1           1               1              1   
9999754              2           2               2              2   
9999759              1           1               1              1   
9999762              1           1               1              1   
9999780              1           1               1              1   
9999789              7           7               7              7   
9999791              1           1               1              1   
9999830              1           1               1              1   
9999838              2           2               2              2   
9999854              1           1               1              1   
9999860              3           3               3              3   
9999862              4           4               4              4   
9999876              1           1               1              1   
9999885              2           2               2              2   
9999890              1           1               1              1   
9999896              2           2               2              2   
9999910              1           1               1              1   
9999917              4           4               4              4   
9999937              2           2               2              2   
999994               1           1               1              1   
9999945              1           1               1              1   
9999948              1           1               1              1   
9999960              1           1               1              1   
9999961    

In [115]:
sum(reviews.verified_purchase)

5032988

### What is the structure of your dataset?

> The cleaned dataset contains 6,900,562 unique reviews with 14 features (customer_id, review_id, product_id, product_parent, product_title, product_category, star_rating, helpful_votes, total_votes, vine, verified_purchase, review_headline, review_body, review_date).    
Most of the columns are objects. Some are unique identifiers to either a product, category or the review (customer_id, review_id, product_id); some are strings relating to the the product such as the name of the items, its category or the review contents (product_title, product_category, review_headline, review_body).  
The ratings of each review along with the helpfullness of each review are intergers (star_rating, helpful_votes, total_votes).  
Two columns are boolian as they are either Y(yes) or N(no) (vine, verifited_purchase).  
The date of the review is in datetime format (review_date).  


### What are the main features of interest in your dataset?

> There are a few aspects, I am interested in investigating.   
I'm interested in seeing if there is a relationship between vine reviewers and non-vine reviewers - do they make a difference?
I also want to see if certian people are more likly to leave good reviews only, bad reviews only, or do they cover both aspects?  
Is there any relationship between the contents of a review and the number of helpful votes?  
Are helpful votes only on good or bad reviews?
Do certian words in a title make for a more helpful review?  
Is there are relationship between the date and types of reviews given?

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> I am expecting that the review rating will be one of the most prominent features to be used. Regarding the contents of a review and the number of helpful votes, the review title and review body. For this I am expecting that long reviews will be more helpful to people over shorter revews.

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.

> Make sure that, after every plot or related series of plots, that you
include a Markdown cell with comments about what you observed, and what
you plan on investigating next.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!